python/ray/data/datasource/partitioning.py

```diff
@@ -226,14 +240,24 @@ def __call__(self, path: str) -> Dict[str, str]:
 
         Args:
             path: Input file path to parse.
+
         Returns:
             Dictionary mapping directory partition keys to values from the input file
             path. Returns an empty dictionary for unpartitioned files.
         """
         dir_path = self._dir_path_trim_base(path)
         if dir_path is None:
             return {}
-        return self._parser_fn(dir_path)
+        partitions: Dict[str, str] = self._parser_fn(dir_path)
+
+        field_types = dict(self._scheme.field_types)
+        for key, value in partitions.items():
+            field_types.setdefault(key, _infer_data_type(value))
+
+        return {
+            key: _cast_value(value, field_types[key])
+            for key, value in partitions.items()
+        }
```

Comment:
This allows different paths to potentially deduce different partition schemas -- we need to enforce that schema is consistent